pkg/controller/deployment/deployment_controller.go

```diff
@@ -335,8 +357,16 @@ func (dc *DeploymentController) deletePod(obj interface{}) {
 		}
 	}
 	if d := dc.getDeploymentForPod(pod); d != nil && d.Spec.Strategy.Type == extensions.RecreateDeploymentStrategyType {
-		podList, err := dc.listPods(d)
-		if err == nil && len(podList.Items) == 0 {
+		// Sync if this Deployment now has no more Pods.
+		rsList, err := dc.getReplicaSetsForDeployment(d)
```

Comment:
Technically and based on the api doc for the deployment selector, we should be able to list all pods managed by a deployment simply by using the deployment selector. Unfortunately, we don't respect this in many places. Mind opening a separate issue to make sure this gets fixed?